### PR TITLE
Move focus from ObserverExpression to ObserverGraph

### PR DESCRIPTION
**Pull request by mdickinson**
This PR does some refactoring to enable better caching possibilities for the observe framework (see #1510).

Overview: in the current `main` branch, an observe expression in the usual mini-language (e.g., `"some_trait.items.updated"`) is first parsed to an `ObserverExpression`, and then that expression is turned into a list of `ObserverGraphs` before those graphs are applied to a particular object with a particular handler, dispatcher, etc. The same mini-language string is always compiled to the same list of `ObserverGraphs`, and the `ObserverGraph` type is treated as immutable. So caching the results of the conversion from mini-language string to `ObserverGraphs` seems like a natural thing to do.

Right now the conversion from `ObserverExpression` to `ObserverGraphs` happens at the time that the observer patterns are applied to objects and listeners. This PR moves things around so that the conversion to `ObserverGraphs` happens instead at parsing time. So instead of dealing primarily with `ObserverExpressions` in `has_traits.py`, we now deal primarily with `ObserverGraphs`.

With this refactor, the new functions `compile_str` and `compile_expr` should be natural targets for applying caching.

## Details of changes

- The internal `_parse_expression` helper function in `has_traits.py` has been replaced with a new `_compile_expression` function which returns a list of `ObserverGraph` objects instead of a list of `ObserverExpression` objects.
- On the other side, there's a new public function `apply_observers` in `traits.observation.api`. This is the equivalent of `observe` from `traits.observation.api`, but it expects `ObserverGraph` objects instead of `ObserverExpressions`. (This more-or-less makes `observe` from `traits.observation.api` redundant, but it's public, so we can't get rid of it easily.)
- There's a new top-level `compile_str` function adjacent to and analogous to the `parse` function: the `parse` function  converts a mini-language string to an `ObserverExpression`, while the `compile_str` function  converts a mini-language string to a list of `ObserverGraph` objects.
- There's also a new `compile_expr` function which  converts an `ObserverExpression` to a list of `ObserverGraphs`.
- `has_traits.py` now uses the `compile_str` / `apply_observers` pairing instead of the `parse` / `observe` pairing. In particular, the `ObserverTraits` dictionary on each `HasTraits` subclass now contains `ObserverGraphs` rather than `ObserverExpressions`.

### CODE REVIEW

File: traits/has_traits.py [hunk 1]
```diff
@@ -330,45 +330,43 @@ def handler(instance, event):
...
-        expressions = expression
-    else:
-        expressions = [expression]
-
-    expressions = [
-        observe_api.parse(expr) if isinstance(expr, str) else expr
-        for expr in expressions
-    ]
-    return expressions
+    expressions = expression if isinstance(expression, list) else [expression]
```
Comment (mdickinson) [inline traits/has_traits.py:361]: This change was gratuitous. Revert.

File: traits/has_traits.py [hunk 2]
```diff
@@ -330,32 +330,33 @@ def handler(instance, event):
...
+
+    Converts a list of strings or ObserverExpressions to a list of
+    ObserverGraphs representing the observation patterns to be applied.
 
     Parameters
     ----------
     expression : str or list or ObserverExpression
         A description of what traits are being observed.
-        If this is a list, each item must be a string or an Expression.
+        If this is a list, each item must be a string or an ObserverExpression.
```
Comment (mdickinson) [inline traits/has_traits.py:353]: Drive-by fix of `Expression` to `ObserverExpression`, for clarity.

**Comment by corranwebster** [comment]
API suggestion: swap `compile` and `compile_from_expr` (perhaps calling the latter `compile_from_str` or something like that).

The `compile` function seems more natural to use as:
```
compile(parse('foo'))
```
or
```
compile(parse('foo').then(match(myfilter)))
```
where when you want to go directly from a string you would do `compile_from_str('foo')`

Or just do the python `compile` thing and accept both strings and intermediate forms.

**Comment by mdickinson** [comment]
> API suggestion: swap `compile` and `compile_from_expr` (perhaps calling the latter `compile_from_str` or something like that).

:+1: I'll rename `compile` to `compile_from_str`. The clash with the builtin was getting annoying, both in IDE syntax highlighting and in picking up the wrong thing in tests when I hadn't imported the right `compile`.

> Or just do the python `compile` thing and accept both strings and intermediate forms.

I considered it, but I'd prefer to avoid DWIM-guessing magic in the low level functions. There's enough of that going on in `_compile_expression`.

**Comment by corranwebster** [comment]
Maybe `compile_str` and `compile_expr`?

I'll stop bikeshedding now.  The general approach is solid.

**Comment by mdickinson** [comment]
> Maybe `compile_str` and `compile_expr`?

Sure. Let's do that.

File: traits/has_traits.py [hunk 3]
```diff
@@ -330,32 +330,33 @@ def handler(instance, event):
...
         handler_getter=handler_getter,
         post_init=False,
     )
 
 
-def _parse_expression(expression):
-    """ Parse a given expression or list of expressions to convert them into
-    ObserveExpression objects if needed.
+def _compile_expression(expression):
+    """ Parse a user-supplied expression or list of expressions.
```
Comment (mdickinson) [inline traits/has_traits.py:344]:
```suggestion
    """ Compile a user-supplied expression or list of expressions.
```

File: traits/has_traits.py [hunk 4]
```diff
@@ -364,11 +365,13 @@ def _parse_expression(expression):
     else:
         expressions = [expression]
 
-    expressions = [
-        observe_api.parse(expr) if isinstance(expr, str) else expr
-        for expr in expressions
-    ]
-    return expressions
+    graphs = []
```
Comment (mdickinson) [inline traits/has_traits.py:368]: Yet again I find myself wishing for a `concat` built-in in Python.

**Review by corranwebster — APPROVED** [review]
This looks good.  The change is small and straightforward.

**Comment by corranwebster** [comment]
Do we need to update the documentation anywhere for this change?

**Comment by rahulporuri** [comment]
I'm very late to the game so it's not clear to me why we need both an `ObserverExpression` and an `ObserverGraph`. If the `ObserverGraph` is what matters in the end, why do we have the intermediate `ObserverExpression`? I looked up the EEP 2 but that didn't provide clarify this question.

**Comment by corranwebster** [comment]
> I'm very late to the game so it's not clear to me why we need both an ObserverExpression and an ObserverGraph. If the ObserverGraph is what matters in the end, why do we have the intermediate ObserverExpression?

The expressions give a programatic way to easily create observation patterns and provide a hook to extend capabilities (eg. new filters, etc).  This was suggested in the EEP in the section beginning "To support the new language, we also want a way to programatically generate pattens as an intermediate form." although the eventual implementation was different.

**Comment by mdickinson** [comment]
Whoops; that last commit should _not_ have gone into this PR. I'll revert, and then merge.

**Comment by mdickinson** [comment]
> Do we need to update the documentation anywhere for this change?

Sorry, I missed this comment earlier. The new functions will be documented in the API docs, but there's not currently any mention of them in the user guide. I think that's fine - they're really only made available in `traits.observation.api` so that `has_traits.py` can use them. I can't yet think of cases where I'd recommend that users user them directly.
